Add Variant class to optimise URL generation performance in request

stevehodgkiss opened the pull request:
# Description

This builds on changes from a separate PR: https://github.com/imgix/imgix-rb/pull/103

Rather than escape the same query string on each request, it can be pre-generated on application start. This removes any need to escape query strings in requests. A similar approach could be taken for `to_srcset` with a `SrcSet` class.

An application might be defining Imgix options in a hash like this:

```ruby
IMGIX_OPTIONS = {
  type_1: { auto: 'compress,format', fit: 'crop', crop: 'top', w: 200, h: 200 },
  type_2: { auto: 'compress,format', fit: 'crop', crop: 'top', w: 400, h: 400 },
}
```

The application could change the definitions to:

```ruby
IMGIX_OPTIONS = {
  type_1: Imgix::Variant.new(auto: 'compress,format', fit: 'crop', crop: 'top', w: 200, h: 200),
  type_2: Imgix::Variant.new(auto: 'compress,format', fit: 'crop', crop: 'top', w: 400, h: 400),
}
```

And then pass the variant to `Imgix::Path#to_url`:

```ruby
client.path('/image.jpg').to_url(variant: IMGIX_OPTIONS[:type_1])
```

Bench output:

```
❯ bundle exec ruby script/bench_to_url.rb
Warming up --------------------------------------
  Imgix::Path#to_url     2.772k i/100ms
Imgix::Path#to_url with variant
                        25.965k i/100ms
Calculating -------------------------------------
  Imgix::Path#to_url     28.021k (± 0.9%) i/s -    141.372k in   5.045620s
Imgix::Path#to_url with variant
                        262.187k (± 1.2%) i/s -      1.324M in   5.051381s

Comparison:
Imgix::Path#to_url with variant:   262186.7 i/s
  Imgix::Path#to_url:    28021.2 i/s - 9.36x  (± 0.00) slower
```

## Checklist: New Feature

- [ ] Each commit follows the [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) format: e.g. `chore(readme): fixed typo`. See the end of this file for more information.
- [ ] Any breaking changes are specified on the commit on which they are introduced with `BREAKING CHANGE` in the body of the commit.
- [ ] If this is a big feature with breaking changes, consider [opening an issue](https://github.com/imgix/imgix-rb/issues/new?labels=&template=feature_request.md&title=) to discuss first. This is completely up to you, but please keep in mind that your PR might not be accepted.
- [ ] Run unit tests to ensure all existing tests are still passing
- [ ] Add new passing unit tests to cover the code introduced by your PR
- [ ] Update the readme
- [ ] Add some [steps](#steps-to-test) so we can test your cool new feature!

stevehodgkiss committed:
test: add benchmark script for Imgix::Path#to_url

Output:

```
❯ bundle exec ruby script/bench_to_url.rb
Warming up --------------------------------------
  Imgix::Path#to_url     1.683k i/100ms
Calculating -------------------------------------
  Imgix::Path#to_url     16.886k (± 0.6%) i/s -     85.833k in   5.083431s
```

stevehodgkiss committed:
perf: avoid calculating path_and_params twice

Benchmark output:

```
❯ bundle exec ruby script/bench_to_url.rb
Warming up --------------------------------------
  Imgix::Path#to_url     2.731k i/100ms
Calculating -------------------------------------
  Imgix::Path#to_url     27.278k (± 0.7%) i/s -    136.550k in   5.006213s
```

sherwinski commented:
Hey @stevehodgkiss, thanks for opening for the PR. I'm interested in reviewing this but it looks like there's a file missing (`imgix/variant`) that still needs to be checked in. That is unless this PR is still a WIP, in which case feel free to ping me when you're ready 👍

stevehodgkiss commented:
Hi @sherwinski, whoops, that's fixed now. Ready for review - interested in your thoughts on the change. A similar approach could be taken to improve `to_srcset` performance.

stevehodgkiss committed:
perf: add Variant class to pre-escape query strings on boot

This avoids needing to do the same thing each request. Rather that
define variants as a hash of options, they can be defined as a Variant
and they won't need to be escaped on each request.

Bench output:

```
❯ bundle exec ruby script/bench_to_url.rb
Warming up --------------------------------------
  Imgix::Path#to_url     2.772k i/100ms
Imgix::Path#to_url with variant
                        25.965k i/100ms
Calculating -------------------------------------
  Imgix::Path#to_url     28.021k (± 0.9%) i/s -    141.372k in   5.045620s
Imgix::Path#to_url with variant
                        262.187k (± 1.2%) i/s -      1.324M in   5.051381s

Comparison:
Imgix::Path#to_url with variant:   262186.7 i/s
  Imgix::Path#to_url:    28021.2 i/s - 9.36x  (± 0.00) slower
```

sherwinski commented:
Hey @stevehodgkiss,
First off, I wanted to say thanks again for opening this PR and putting so much thought into improving this library in this way. Our team is always looking for ways to improve performance as well as usability of our SDKs; and this PR has pushed us to think more deeply about how else we can prevent repeat work. Having said that, we are opting to not accept this change at this time, and I’d like to explain why.
After discussing it further with the team, we agreed that this is a feature that should be implemented across our [entire kit](https://docs.imgix.com/libraries#client-libraries) of open source libraries, spanning approximately 10 different languages. As a result, this requires extra work before release: ensuring cross-kit consistency, considering language-specific idioms, and investigating for potential edge cases. For this reason, we’d like to put more time into designing this feature internally before implementing it across the board.
